### ISSUE TITLE
License == ?

### ISSUE THREAD

**Issue by guillaume-chevalier** [ternaus/kaggle_dstl_submission#1]
Hi, I would like to know under which license is your work available. GitHub's default license is "all rights reserved" for public repositories, such as this one. 

I have read the following, about winners' obligations of publishing the solution: https://www.kaggle.com/c/dstl-satellite-imagery-feature-detection/rules
I am not exactly sure of what type of competition this is, so I am not sure under which license you have published your solution according to the rules. I would like to know it. Adding a license to your repository is simple: 
https://help.github.com/articles/adding-a-license-to-a-repository/
Just be sure that in the template you pick to add your names. 

By the way, the following piece of code is licensed Apache 2.0: https://www.kaggle.com/ceperaang/lb-0-42-ultimate-full-solution-run-on-your-hw/code/code
And it also use other Apache 2.0 Licensed content. But the license for that kernel may not be the same as for this GitHub repository. 

Many thanks!

**Comment by johnny-salz** [comment]
Feel free to use as you want! Added MIT's licence.

**Comment by guillaume-chevalier** [comment]
Thanks! If you are interested in seeing what I have done with your code, see my post here: 
https://vooban.com/en/tips-articles-geek-stuff/satellite-image-segmentation-workflow-with-u-net/

By the way, I am wondering if you have used automatic hyperparameter optimization search (such as using Hyperopt or a grid search), or if you did that incrementally and manually with potentially less computing power.

I also wonder about how did you split the labelled data into a training/validation/test set, so as to find your best architecture. On my side, I used a test set as being the first fifth of the 25 images at the left (in other words, the first left column caused no class imbalance as a train/test split): 
![](https://vooban.com/wp-content/uploads/2017/08/full-dstl-training-set.jpg)

I would be interested in having your output on that, if that does not bother you / if you have time for that :) Thanks again!

**Comment by ternaus** [comment]
I loved your post and amount of work that you put into it is very impressive.

We did not use HypertOpt or gridsearch for this particular problem. It is expensive and we just did not have enough hardware. I used HyperOpt in other problems. It is pretty nice and works better than BayesianOpt

1. All Unet modifications architectures work similarly. The only important difference was a BatchNorm that increased convergence a lot. A different number of filters, different activation functions, a different number of blocks - all nearly the same. One that we used worked well for me in the UltraSound Nerve segmentation challenge.
 
2. To my knowledge, noone figured out how to do reliable validation in this problem. Train, Public and Private test are small and very different. We tried picking several images, like what you described, we tried to split each image into train and test parts - no success. Both these variations are able to catch big improvements, but nothing is able to catch small. => in this problem we took advantage of the fact that overfitting Unet is hard and used visual inspection (plotting images and predicted masks side by side) and feedback from Public LB which was unreliable.

3. In your post you mentioned Tiramisu. - We and other teams tried it, no success. Same performance as Unet but converges much significantly slower. Most likely one will be able to make Tiramisu work well, but probably not on this dataset but something much bigger.

How I would try to imporve our solution if I did it right now?

1. After that challenge I used mxnet, but right now I am using PyTorch and I really like it. => my first step would be to transfer everything into Pytorch, especially because Konstantin Lophin already has really nice Unet implementation: https://github.com/lopuhin/mapillary-vistas-2017 . You may also find his code for the DSTL problem there (his team finished 5th). https://github.com/lopuhin/kaggle-dstl 
Reason, why I would use PyTorch, is because I have multiple GPU at home and PyTorch allows to parallelize your batch without pain, that you get in TensorFlow => instead of 112x112x128 I would be able to use, say 224x224x128 or 256x256x64 which, I believe would help for some classes.

2. Another thing that I had in mind, but we did not have time to implement is different input scale for different images. I believe downscaling input images and feeding to the same network would help trees and crops. Something that worked well for Kyle, who finished 1st.

3. And of course car classes. A lot can be done there, even taking into account far from the perfect labeling quality. Sergey and I got a lot of practice in this type of the task in a https://www.datasciencechallenge.org/challenges/1/safe-passage where one needed to find cars from aerial images. I finished second, and I believe some knowledge in that problem may be transferred to this type of the task.

Again, I loved your post and reposted it where I could :)

**Comment by guillaume-chevalier** [comment]
Thanks for sharing! I edited the post to include recommendations on the amount of data if using the Tiramisu. Also, I didn't know that the 1st place winner used different zoom levels as data augmentation. 

PyTorch seems very interesting to me too, and especially for its dynamic computation graph (random fact: I have just started learning PyTorch yesterday evening hehe, that's motivating today to see your thoughts on it).

Thanks again :)

**Comment by guillaume-chevalier** [comment]
By the way, I have also added a link in the blog post to some code I have made for making the prediction perfectly smooth (D4 Dihedral Group for predictions + smooth merging of overlapped image patches):
https://github.com/Vooban/Smoothly-Blend-Image-Patches
That GitHub repository I just linked might soon be converted into another blog post on the website.
